LoginServer/PipeServer.cs

```diff
@@ -0,0 +1,558 @@
+
+                        do
+                        {
+                            int numBytes = client.stream.Read(buffer, 0, Math.Min(totalSize - bytesRead, BUFFER_SIZE));
+
+                            ms.Write(buffer, 0, numBytes);
+
+                            bytesRead += numBytes;
+
+                        } while (bytesRead < totalSize);
+
+                    }
+                    catch
+                    {
+                        Console.Out.WriteLine("Server read break 2");
+                        break;
+                    }
+                    Console.Out.WriteLine();
+                    Console.Out.WriteLine("Bytes read before client disconnected." + bytesRead);
+                    // Client has disconnected.
+                    if (bytesRead == 0)
+                    {
+                        Console.Out.WriteLine("Server read break 3");
+                        break;
+                    }
+                    Console.Out.WriteLine("Bytes read after client disconnected." + bytesRead);
+                    Console.Out.WriteLine();
+
+                    if (MessageRecieved != null)
+                    {
+                        Console.Out.WriteLine();
+                        string str = Utility.ConvertToString(ms.ToArray());
+                        Console.Out.WriteLine("Now printing the MessageRecieved.");
+                        Console.Out.WriteLine(str);
+                        Console.Out.WriteLine("After the MessageRecieved from the client: " + client.id);
+                        Console.Out.WriteLine("Total clients: " + clients.Count);
+                        Console.Out.WriteLine("Clients Handle: " + client.handle.ToString());
+                        Console.Out.WriteLine();
+
+                        if (client.isLoggedIn)
+                        {
+                            MessageRecieved(ms.ToArray());
+                        }
+                        else
+                        {
+                            Console.Out.WriteLine(Thread.CurrentThread.ManagedThreadId + " in Read");
+                            ValidateClient(ms.ToArray(), client);
+                        }
+                    }
+                }
+            }
+
+            // The clients must be locked, otherwise "stream.Close()"
+            // could be called while SendMessage(byte[]) is being called on another thread.
+            // This would lead to an ID error.
+            lock (clients)
+            {
+                // Clean up the resources
+                DisconnectNamedPipe(client.handle);
+                client.stream.Close();
+                client.handle.Close();
+
+                // remove the client from the list of clients.
+                clients.Remove(client);
+            }
+
+            // invoke the event, a client disconnected.
+            if (ClientDisconnected != null)
+            {
+                ClientDisconnected();
```

Comment:
Not a bug, just a thought:
In a real high traffic server.
You wouldn't fire the event inside the closing thread.  10000's of players could all disconnect at once, and sending that many messages, would be fatal.